### PR TITLE
NEWEBODAC-27: Create user role management page

### PR DESCRIPTION
**Pull request by mariaogryczak**
For now, could you just check what is wrong with connecting the components, that the added role does not appear on the role list?

### CODE REVIEW

File: src/main/webapp/js/actions/types.js [hunk 1]
```diff
@@ -16,3 +16,20 @@ export const SAVE_FIELD_CONFIG_ORDER = 'save_field_config_order';
 
 export const CHANGE_FIELD_VISIBILITY = 'change_field_visibility';
 export const CHANGE_FIELD_ORDER = 'change_field_order';
+export const FETCH_VACCINEE_CONFIG = 'fetch_vaccinee_config';
```
Comment (pmuchowski) [inline src/main/webapp/js/actions/types.js]: remove all of those types, there are not use anywhere

File: src/main/java/org/motechproject/newebodac/constants/DefaultPermissions.java [hunk 2]
```diff
@@ -3,7 +3,60 @@
 @SuppressWarnings({"PMD.AvoidDuplicateLiterals", "PMD.ClassNamingConventions"})
 public class DefaultPermissions {
 
+  public static final String VACCINEE_READ = "ROLE_VACCINEE_READ";
+  public static final String VACCINEE_WRITE = "ROLE_VACCINEE_WRITE";
+  public static final String INCHARGE_READ = "ROLE_INCHARGE_READ";
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/newebodac/constants/DefaultPermissions.java]: there are no incharges in this project, remove all unused permissions

File: src/main/webapp/js/reducers/entity-reducer.js [hunk 3]
```diff
@@ -22,6 +24,7 @@ export default (state = initialState, action) => {
     case FETCH_ENTITY:
       return update(state, {
         [entityType]: { $set: _.keyBy(payload.data, 'id') },
+        role: action.payload.data,
```
Comment (pmuchowski) [inline src/main/webapp/js/reducers/entity-reducer.js]: the entityType will be 'role' in this case, so remove this

File: src/main/webapp/js/reducers/field-config-reducer.js [hunk 4]
```diff
@@ -16,6 +16,8 @@ const initialState = {
   keyCommunityPerson: {},
   site: {},
   visit: {},
+  role: {},
```
Comment (pmuchowski) [inline src/main/webapp/js/reducers/field-config-reducer.js]: the field config for roles and permissions is not fetched from the backend, but it's a constant, so remove this

File: src/main/webapp/js/components/navbar.jsx [hunk 5]
```diff
@@ -75,14 +76,15 @@ class Navbar extends Component {
...
-                    <RoutePrivate path="/roles" component={Roles} />
                     <RoutePrivate path="/users" component={Users} />
+                    <RoutePrivate path="/role" component={Roles} />
                     <RoutePrivate path="/enrollmentGroup" component={Enrollment} />
                     <RoutePrivate path="/messaging" component={Messaging} />
                     <RoutePrivate path="/languages" component={Languages} />
                     <RoutePrivate path="/sites" component={Sites} />
                     <RoutePrivate path="/vaccineeFields" component={VaccineeFields} />
                     <RoutePrivate path="/keyCommunityPersonFields" component={KeyCommunityPersonFields} />
+                    <RoutePrivate path="/roleFields" component={RoleFields} />
```
Comment (pmuchowski) [inline src/main/webapp/js/components/navbar.jsx]: the field config for roles is not editable, so remove this

File: src/main/webapp/js/components/sidebar.jsx [hunk 6]
```diff
@@ -253,6 +253,14 @@ class SideBar extends Component {
             </Link>
           </div>
         </li>
+        <li className="border-none">
```
Comment (pmuchowski) [inline src/main/webapp/js/components/sidebar.jsx]: remove this, the field roles field config is not editable

File: src/main/webapp/js/components/fields/role-fields.jsx [hunk 7]
```diff
@@ -0,0 +1,14 @@
+import React from 'react';
+
+import FieldConfigPage from './field-config-page';
+import { ROLE_ENTITY } from '../../utils/entity-types';
+
+const RoleFields = () => (
```
Comment (pmuchowski) [inline src/main/webapp/js/components/fields/role-fields.jsx]: remove this, role field config is not editable

File: src/main/webapp/css/main.scss [hunk 8]
```diff
@@ -555,3 +555,18 @@ select.form-control {
   font-weight: bold;
   text-align: center;
 }
+
+.role-buttons {
```
Comment (pmuchowski) [inline src/main/webapp/css/main.scss]: are those classes used anywhere, if not remove it

File: src/main/webapp/js/utils/entity-types.js [hunk 9]
```diff
@@ -4,3 +4,4 @@ export const SITE_ENTITY = 'site';
 export const KEY_COMMUNITY_PERSON_ENTITY = 'keyCommunityPerson';
 export const GROUP_ENTITY = 'group';
 export const LANGUAGE_ENTITY = 'language';
+export const ROLE_ENTITY = 'role';
```
Comment (pmuchowski) [inline src/main/webapp/js/utils/entity-types.js:7]: you will need to add entity for the permission too

File: src/main/java/org/motechproject/newebodac/web/RoleController.java [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+      @RequestBody @Valid RoleDto roleDto) {
+    return roleService.updateFromDto(id, roleDto);
+  }
+
+  @RequestMapping(value = "/role/delete", method = RequestMethod.DELETE)
+  @ResponseStatus(HttpStatus.OK)
+  @ResponseBody
+  public void delete(@PathVariable("id") UUID id) {
+    roleService.delete(id);
+  }
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/newebodac/web/RoleController.java:79]: you will need to add a separate endpoint to fetch all permissions

File: src/main/java/org/motechproject/newebodac/web/RoleController.java [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+  public RoleDto findById(@PathVariable("id") UUID id) {
+    return roleService.findById(id);
+  }
+
+  /**
+   * Creates role from given dto and saves it into the database.
+   * @param roleDto Dto of created role.
+   * @return Dto od created role.
+   */
+  @RequestMapping(value = "/role/create", method = RequestMethod.POST)
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/newebodac/web/RoleController.java]: remove '/create' from the url

File: src/main/java/org/motechproject/newebodac/web/RoleController.java [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+
+  @RequestMapping(value = "/role/{id}", method = RequestMethod.PUT)
+  @ResponseStatus(HttpStatus.OK)
+  @ResponseBody
+  public RoleDto update(@PathVariable("id") UUID id,
+      @RequestBody @Valid RoleDto roleDto) {
+    return roleService.updateFromDto(id, roleDto);
+  }
+
+  @RequestMapping(value = "/role/delete", method = RequestMethod.DELETE)
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/newebodac/web/RoleController.java]: remove '/delete' from the url

File: src/main/java/org/motechproject/newebodac/service/RoleService.java [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+
+@Service
+public class RoleService {
+
+  private static final RoleMapper MAPPER = RoleMapper.INSTANCE;
+  private static final PermissionMapper PERMISSION_MAPPER = PermissionMapper.INSTANCE;
+
+  @Autowired
+  private RoleRepository roleRepository;
+  private PermissionRepository permissionRepository;
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/newebodac/service/RoleService.java:26]: you will need to add `@Autowired` on this field too

File: src/main/java/org/motechproject/newebodac/service/RoleService.java [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+    return MAPPER.toDto(roleRepository.save(MAPPER.fromDto(roleDto)));
+  }
+
+  /**
+   * Updates data from dto to role object, saves it and returns dto of it.
+   * @param id ID of object to update.
+   * @param roleDto Dto of role to update.
+   * @return Dto of of updated role
+   */
+  public RoleDto updateFromDto(UUID id, RoleDto roleDto) {
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/newebodac/service/RoleService.java]: rename this to 'update'

File: src/main/java/org/motechproject/newebodac/mapper/RoleMapper.java [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+import org.motechproject.newebodac.domain.security.UserRole;
+import org.motechproject.newebodac.dto.RoleDto;
+
+@Mapper(uses = { UuidMapper.class },
+    unmappedTargetPolicy = ReportingPolicy.IGNORE)
+public interface RoleMapper extends EntityMapper<RoleDto, UserRole> {
+
+  RoleMapper INSTANCE = Mappers.getMapper(RoleMapper.class);
+
+  RoleDto toDto(UserRole role);
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/newebodac/mapper/RoleMapper.java]: if you're not changing any mappers, you don't need to implement this (there is default implementation in EntityMapper), so remove it

File: src/main/java/org/motechproject/newebodac/mapper/RoleMapper.java [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+
+  @Mappings({
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/newebodac/mapper/RoleMapper.java]: there is not need for this, this was removed in newer versions of mapstruct

File: src/main/java/org/motechproject/newebodac/mapper/RoleMapper.java [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+      @Mapping(target = "id", ignore = true),
+      @Mapping(target = "readonly", ignore = true)
+  })
+  UserRole fromDto(RoleDto roleDto);
+
+  List<RoleDto> toDtos(Iterable<UserRole> role);
+
+  @Mappings({
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/newebodac/mapper/RoleMapper.java]: as above

File: src/main/java/org/motechproject/newebodac/mapper/RoleMapper.java [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+
+  RoleDto toDto(UserRole role);
+
+  @Mappings({
+      @Mapping(target = "id", ignore = true),
+      @Mapping(target = "readonly", ignore = true)
+  })
+  UserRole fromDto(RoleDto roleDto);
+
+  List<RoleDto> toDtos(Iterable<UserRole> role);
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/newebodac/mapper/RoleMapper.java]: you can remove it

File: src/main/java/org/motechproject/newebodac/dto/RoleDto.java [hunk 19]
```diff
@@ -0,0 +1,20 @@
...
+import java.util.Set;
+import lombok.Getter;
+import lombok.Setter;
+
+
+@Getter
+@Setter
+public class RoleDto extends BaseDto {
+
+  private String roleId;
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/newebodac/dto/RoleDto.java]: rename this to id

**Review by pmuchowski — CHANGES_REQUESTED** [review]

**Comment by mariaogryczak** [comment]
@pmuchowski 
Could you check what I did so far?
Could you give me any suggestion how to connect Role Entity with Permission Entity (I suppose I need to use relatedEntity field)? And how to connect Default Permissions with all of this?

File: src/main/java/org/motechproject/newebodac/mapper/RoleMapper.java [hunk 20]
```diff
@@ -0,0 +1,17 @@
...
+import org.motechproject.newebodac.domain.security.UserRole;
+import org.motechproject.newebodac.dto.RoleDto;
+
+@Mapper(uses = { UuidMapper.class },
+    unmappedTargetPolicy = ReportingPolicy.IGNORE)
+public interface RoleMapper extends EntityMapper<RoleDto, UserRole> {
+
+  RoleMapper INSTANCE = Mappers.getMapper(RoleMapper.class);
+
+  void update(RoleDto roleDto, @MappingTarget UserRole role);
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/newebodac/mapper/RoleMapper.java:22]: add ignore for id, createDate, updateDate and readonly fields

File: src/main/java/org/motechproject/newebodac/web/RoleController.java [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+  @ResponseBody
+  public List<RoleDto> getAll() {
+    return roleService.getAll();
+  }
+
+  /**
+   * Returns List of user permissions Dtos loaded from the database.
+   * @return List of user permissions Dtos
+   */
+  @RequestMapping(value = "/role/permission", method = RequestMethod.GET)
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/newebodac/web/RoleController.java]: change the url to '/permission'

File: src/main/java/org/motechproject/newebodac/dto/RoleDto.java [hunk 22]
```diff
@@ -0,0 +1,22 @@
...
+
+import java.util.HashSet;
+import java.util.Set;
+import lombok.Getter;
+import lombok.Setter;
+
+
+@Getter
+@Setter
+public class RoleDto {
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/newebodac/dto/RoleDto.java]: it should extend the BaseDto

File: src/main/java/org/motechproject/newebodac/dto/PermissionDto.java [hunk 23]
```diff
@@ -0,0 +1,17 @@
+package org.motechproject.newebodac.dto;
+
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+public class PermissionDto {
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/newebodac/dto/PermissionDto.java]: it should extend the BaseDto

File: src/main/java/org/motechproject/newebodac/dto/RoleDto.java [hunk 24]
```diff
@@ -0,0 +1,22 @@
...
+
+  private String id;
+
+  private String name;
+
+  private Boolean readonly = false;
+
+  private Set<ExtraFieldDto> extraFields;
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/newebodac/dto/RoleDto.java]: remove this

File: src/main/java/org/motechproject/newebodac/dto/RoleDto.java [hunk 25]
```diff
@@ -0,0 +1,22 @@
...
+
+  private Set<PermissionDto> permissions = new HashSet<>();
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/newebodac/dto/RoleDto.java]: change it to `Set<UUID>` and store only ids of permissions here

File: src/main/java/org/motechproject/newebodac/domain/security/UserRole.java [hunk 26]
```diff
@@ -10,18 +11,18 @@
...
 import org.motechproject.newebodac.domain.BaseEntity;
 
 @Getter
 @Setter
 @Entity
 @Table(name = "user_role")
+@NoArgsConstructor
 public class UserRole extends BaseEntity {
-
-  @NotBlank
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/newebodac/domain/security/UserRole.java:24]: you should not remove this

File: src/main/java/org/motechproject/newebodac/mapper/RoleMapper.java [hunk 27]
```diff
@@ -0,0 +1,17 @@
...
+import org.mapstruct.factory.Mappers;
+import org.motechproject.newebodac.domain.security.UserRole;
+import org.motechproject.newebodac.dto.RoleDto;
+
+@Mapper(uses = { UuidMapper.class },
+    unmappedTargetPolicy = ReportingPolicy.IGNORE)
+public interface RoleMapper extends EntityMapper<RoleDto, UserRole> {
+
+  RoleMapper INSTANCE = Mappers.getMapper(RoleMapper.class);
+
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/newebodac/mapper/RoleMapper.java:17]: you will need to add new methods here to map UUID to UserPermission and other way around

**Review by pmuchowski — CHANGES_REQUESTED** [review]

File: src/main/java/org/motechproject/newebodac/mapper/RoleMapper.java [hunk 28]
```diff
@@ -0,0 +1,28 @@
...
+  @Mapping(target = "id", ignore = true)
+  @Mapping(target = "createDate", ignore = true)
+  @Mapping(target = "updateDate", ignore = true)
+  @Mapping(target = "readonly", ignore = true)
+  void update(RoleDto roleDto, @MappingTarget UserRole role);
+
+  default UserPermission toUserPermission(UUID id) {
+    return new UserPermission();
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/newebodac/mapper/RoleMapper.java]: you need to set the id for the UserPermission, also add another method to map UserPermission to UUID

File: src/main/java/org/motechproject/newebodac/mapper/PermissionMapper.java [hunk 29]
```diff
@@ -0,0 +1,18 @@
+package org.motechproject.newebodac.mapper;
+
+import java.util.List;
+import org.mapstruct.Mapper;
+import org.mapstruct.ReportingPolicy;
+import org.mapstruct.factory.Mappers;
+import org.motechproject.newebodac.domain.security.UserPermission;
+import org.motechproject.newebodac.dto.PermissionDto;
+
+@Mapper(uses = { UuidMapper.class },
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/newebodac/mapper/PermissionMapper.java]: remove the UuidMapper

File: src/main/java/org/motechproject/newebodac/mapper/PermissionMapper.java [hunk 30]
```diff
@@ -0,0 +1,18 @@
...
+    unmappedTargetPolicy = ReportingPolicy.IGNORE)
+public interface PermissionMapper {
+
+  PermissionMapper INSTANCE = Mappers.getMapper(PermissionMapper.class);
+
+  List<PermissionDto> toDtos(Iterable<UserPermission> id);
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/newebodac/mapper/PermissionMapper.java]: rename the argument, it should be permissions or something similar, not id

File: src/main/java/org/motechproject/newebodac/mapper/RoleMapper.java [hunk 31]
```diff
@@ -0,0 +1,28 @@
...
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+import org.mapstruct.MappingTarget;
+import org.mapstruct.ReportingPolicy;
+import org.mapstruct.factory.Mappers;
+import org.motechproject.newebodac.domain.security.UserPermission;
+import org.motechproject.newebodac.domain.security.UserRole;
+import org.motechproject.newebodac.dto.RoleDto;
+
+@Mapper(uses = { UuidMapper.class },
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/newebodac/mapper/RoleMapper.java]: remove the UuidMapper

File: src/main/java/org/motechproject/newebodac/dto/RoleDto.java [hunk 32]
```diff
@@ -0,0 +1,19 @@
...
+
+@Getter
+@Setter
+public class RoleDto extends BaseDto {
+
+  private String name;
+
+  private Boolean readonly = false;
+
+  private Set<UUID> permissionId = new HashSet<>();
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/newebodac/dto/RoleDto.java]: this should be named permissionIds

File: src/main/webapp/js/components/fields/field-config-modal.jsx [hunk 33]
```diff
@@ -12,7 +12,7 @@ import { createFieldConfig, deleteFieldConfig, saveFieldConfig } from '../../act
 
 Modal.setAppElement(document.getElementById('root'));
 
-const defaultConfig = {
+export const defaultConfig = {
```
Comment (pmuchowski) [inline src/main/webapp/js/components/fields/field-config-modal.jsx]: it's not used anywhere, so remove the export

File: src/main/webapp/js/selectors/index.js [hunk 34]
```diff
@@ -1,7 +1,7 @@
 import _ from 'lodash';
 import { createSelector } from 'reselect';
 
-const getFieldConfigByEntity = (state, { entityType }) => state.fieldConfig[entityType];
+export const getFieldConfigByEntity = (state, { entityType }) => state.fieldConfig[entityType];
```
Comment (pmuchowski) [inline src/main/webapp/js/selectors/index.js]: it's not used anywhere, so remove the export

File: src/main/webapp/js/components/user-managment/roles.jsx [hunk 35]
```diff
@@ -1,14 +1,75 @@
...
+  render() {
+    return (
+      <div className="container-fluid">
+        <h1>
+          Roles
+        </h1>
+        <EntityTable
+          entityType={ROLE_ENTITY}
+          fieldConfig={defaultRoleConfig}
+          relatedEntity={PERMISSION_ENTITY}
```
Comment (pmuchowski) [inline src/main/webapp/js/components/user-managment/roles.jsx]: remove this, relatedEntity is not one of the EntityTable properties

File: src/main/webapp/js/components/user-managment/roles.jsx [hunk 36]
```diff
@@ -1,14 +1,75 @@
...
+  displayName: 'Permissions',
+  name: 'user_role_permissions',
+  relatedEntity: PERMISSION_ENTITY,
+  relatedField: null,
+  format: '',
+},
+];
+
+class Roles extends Component {
+  componentDidMount() {
```
Comment (pmuchowski) [inline src/main/webapp/js/components/user-managment/roles.jsx:10]: you need to fetch the user roles and permissions here, you can use 'fetchEntity' for this

File: src/main/webapp/js/components/user-managment/roles.jsx [hunk 37]
```diff
@@ -1,14 +1,75 @@
...
-
-export default Roles;
+import React, { Component } from 'react';
+import { connect } from 'react-redux';
+import PropTypes from 'prop-types';
+import EntityTable from '../entities/entity-table';
+import { ROLE_ENTITY, PERMISSION_ENTITY } from '../../utils/entity-types';
+
+export const defaultRoleConfig = [{
+  base: false,
```
Comment (pmuchowski) [inline src/main/webapp/js/components/user-managment/roles.jsx]: all of those fields are base fields

File: src/main/webapp/js/components/user-managment/roles.jsx [hunk 38]
```diff
@@ -1,14 +1,75 @@
...
+  editable: true,
```
Comment (pmuchowski) [inline src/main/webapp/js/components/user-managment/roles.jsx]: id is not editable

File: src/main/webapp/js/components/user-managment/roles.jsx [hunk 39]
```diff
@@ -1,14 +1,75 @@
...
+  filterable: true,
+  required: false,
+  hidden: false,
```
Comment (pmuchowski) [inline src/main/webapp/js/components/user-managment/roles.jsx]: id should be hidden

File: src/main/webapp/js/components/user-managment/roles.jsx [hunk 40]
```diff
@@ -1,14 +1,75 @@
...
+  fieldType: null,
```
Comment (pmuchowski) [inline src/main/webapp/js/components/user-managment/roles.jsx]: field type cannot be null, change it to TEXT

File: src/main/webapp/js/components/user-managment/roles.jsx [hunk 41]
```diff
@@ -1,14 +1,75 @@
...
+  relatedEntity: PERMISSION_ENTITY,
+  relatedField: null,
+  format: '',
+}, {
+  base: false,
+  editable: true,
+  filterable: true,
+  required: false,
+  hidden: false,
+  fieldType: null,
```
Comment (pmuchowski) [inline src/main/webapp/js/components/user-managment/roles.jsx]: field type cannot be null, change it to TEXT

File: src/main/webapp/js/components/user-managment/roles.jsx [hunk 42]
```diff
@@ -1,14 +1,75 @@
...
+import { ROLE_ENTITY, PERMISSION_ENTITY } from '../../utils/entity-types';
+
+export const defaultRoleConfig = [{
+  base: false,
+  editable: true,
+  filterable: true,
+  required: false,
+  hidden: false,
+  fieldType: null,
+  entity: null,
```
Comment (pmuchowski) [inline src/main/webapp/js/components/user-managment/roles.jsx]: entity is ROLE_ENTITY

File: src/main/webapp/js/components/user-managment/roles.jsx [hunk 43]
```diff
@@ -1,14 +1,75 @@
...
+  displayName: 'ID',
+  name: 'id',
+  relatedEntity: PERMISSION_ENTITY,
```
Comment (pmuchowski) [inline src/main/webapp/js/components/user-managment/roles.jsx]: this should be null

File: src/main/webapp/js/components/user-managment/roles.jsx [hunk 44]
```diff
@@ -1,14 +1,75 @@
...
+  relatedField: null,
+  format: '',
+}, {
+  base: false,
+  editable: true,
+  filterable: true,
+  required: false,
+  hidden: false,
+  fieldType: null,
+  entity: null,
```
Comment (pmuchowski) [inline src/main/webapp/js/components/user-managment/roles.jsx]: entity is ROLE_ENTITY

File: src/main/webapp/js/components/user-managment/roles.jsx [hunk 45]
```diff
@@ -1,14 +1,75 @@
...
```
Comment (pmuchowski) [inline src/main/webapp/js/components/user-managment/roles.jsx]: entity is ROLE_ENTITY

File: src/main/webapp/js/components/user-managment/roles.jsx [hunk 46]
```diff
@@ -1,14 +1,75 @@
...
+  displayName: 'ID',
+  name: 'id',
+  relatedEntity: PERMISSION_ENTITY,
+  relatedField: null,
+  format: '',
+}, {
+  base: false,
+  editable: true,
+  filterable: true,
+  required: false,
```
Comment (pmuchowski) [inline src/main/webapp/js/components/user-managment/roles.jsx]: name is required

File: src/main/webapp/js/components/user-managment/roles.jsx [hunk 47]
```diff
@@ -1,14 +1,75 @@
...
+  hidden: false,
+  fieldType: null,
+  entity: null,
+  displayName: 'Name',
+  name: 'name',
+  relatedEntity: PERMISSION_ENTITY,
```
Comment (pmuchowski) [inline src/main/webapp/js/components/user-managment/roles.jsx]: this should be null

File: src/main/webapp/js/components/user-managment/roles.jsx [hunk 48]
```diff
@@ -1,14 +1,75 @@
...
+}, {
+  base: false,
+  editable: true,
+  filterable: true,
+  required: false,
+  hidden: false,
+  fieldType: null,
+  entity: null,
+  displayName: 'Permissions',
+  name: 'user_role_permissions',
```
Comment (pmuchowski) [inline src/main/webapp/js/components/user-managment/roles.jsx]: name should be permissionIds

File: src/main/webapp/js/components/user-managment/roles.jsx [hunk 49]
```diff
@@ -1,14 +1,75 @@
...
+  relatedEntity: PERMISSION_ENTITY,
+  relatedField: null,
```
Comment (pmuchowski) [inline src/main/webapp/js/components/user-managment/roles.jsx]: the related field should be 'displayName'

File: src/main/webapp/js/components/user-managment/roles.jsx [hunk 50]
```diff
@@ -1,14 +1,75 @@
...
+  format: '',
+}, {
+  base: false,
+  editable: true,
+  filterable: true,
+  required: false,
+  hidden: false,
+  fieldType: null,
```
Comment (pmuchowski) [inline src/main/webapp/js/components/user-managment/roles.jsx]: field type is 'COLLECTION'

File: src/main/webapp/js/components/user-managment/roles.jsx [hunk 51]
```diff
@@ -1,14 +1,75 @@
...
-);
-
-export default Roles;
+import React, { Component } from 'react';
+import { connect } from 'react-redux';
+import PropTypes from 'prop-types';
+import EntityTable from '../entities/entity-table';
+import { ROLE_ENTITY, PERMISSION_ENTITY } from '../../utils/entity-types';
+
+export const defaultRoleConfig = [{
```
Comment (pmuchowski) [inline src/main/webapp/js/components/user-managment/roles.jsx]: rename this to roleFieldConfig and create a new file named 'field-configs.js' (in utils) and move this config there

File: src/main/java/org/motechproject/newebodac/constants/DefaultPermissions.java [hunk 52]
```diff
@@ -4,6 +4,19 @@
...
+  public static final String DISPLAY_REPORTS = "ROLE_DISPLAY_REPORTS";
+  public static final String GROUP_READ = "ROLE_GROUP_READ";
+  public static final String GROUP_WRITE = "ROLE_GROUP_WRITE";
+  public static final String HAS_VACCINEE_READ_ROLE = "hasRole('" + VACCINEE_READ + "')";
+  public static final String HAS_VACCINEE_WRITE_ROLE = "hasRole('" + VACCINEE_WRITE + "')";
   public static final String HAS_ADMIN_ROLE = "hasRole('" + ADMIN + "')";
+  public static final String HAS_MANAGE_USERS_ROLE = "hasRole('" + MANAGE_USERS + "')";
+  public static final String HAS_DISPLAY_REPORTS_ROLE = "hasRole('" + DISPLAY_REPORTS + "')";
+  public static final String HAS_GROUP_READ_ROLE = "hasRole('" + GROUP_READ + "')";
+  public static final String HAS_GROUP_READ_OR_VACCINEE_WRITE =
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/newebodac/constants/DefaultPermissions.java]: are you using this anywhere?

**Review by pmuchowski — CHANGES_REQUESTED** [review]

File: src/main/java/org/motechproject/newebodac/mapper/RoleMapper.java [hunk 53]
```diff
@@ -0,0 +1,33 @@
...
+
+  RoleMapper INSTANCE = Mappers.getMapper(RoleMapper.class);
+
+  @Mapping(target = "id", ignore = true)
+  @Mapping(target = "createDate", ignore = true)
+  @Mapping(target = "updateDate", ignore = true)
+  @Mapping(target = "readonly", ignore = true)
+  void update(RoleDto roleDto, @MappingTarget UserRole role);
+
+  default Set<UserPermission> toUserPermission(UUID id) {
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/newebodac/mapper/RoleMapper.java]: this should not be a Set, just one UserPermission, return `new UserPermission(id)` here

File: src/main/java/org/motechproject/newebodac/mapper/RoleMapper.java [hunk 54]
```diff
@@ -0,0 +1,33 @@
...
+    return new HashSet<>();
+  }
+
+  default Set<UUID> toUuid(UserPermission id) {
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/newebodac/mapper/RoleMapper.java]: change the parameter name to userPermission and return UserPermission id here (single id not Set of ids)

File: src/main/webapp/js/components/navbar.jsx [hunk 55]
```diff
@@ -83,6 +84,7 @@ class Navbar extends Component {
                     <RoutePrivate path="/messageCampaign" component={MessageCampaign} />
                     <RoutePrivate path="/viewEntity/:entityType" component={ViewEntity} />
                     <RoutePrivate path="/entityEdit/:entityType" component={EditEntityPage} />
+                    <RoutePrivate path="/entityEdit/role/:id" component={RoleEditPage} />
```
Comment (pmuchowski) [inline src/main/webapp/js/components/navbar.jsx]: it's fine, but would be better if you used ROLE_ENTITY const here

File: src/main/webapp/js/components/user-managment/roles.jsx [hunk 56]
```diff
@@ -1,14 +1,37 @@
...
+      </div>
+    );
+  }
+}
+
+export default connect(null, { fetchEntity })(Roles);
+
+Roles.propTypes = {
+  fetchEntity: PropTypes.func.isRequired,
+  history: PropTypes.shape({
```
Comment (pmuchowski) [inline src/main/webapp/js/components/user-managment/roles.jsx]: you don't need the history prop here

File: src/main/webapp/js/components/user-managment/roles.jsx [hunk 57]
```diff
@@ -1,14 +1,37 @@
...
+import { connect } from 'react-redux';
+import PropTypes from 'prop-types';
+import { fetchEntity } from '../../actions';
+import EntityTable from '../entities/entity-table';
+import { ROLE_ENTITY, PERMISSION_ENTITY } from '../../utils/entity-types';
+import roleFieldConfig from '../../utils/field-configs';
+
+class Roles extends Component {
+  componentDidMount() {
+    this.props.fetchEntity(ROLE_ENTITY);
```
Comment (pmuchowski) [inline src/main/webapp/js/components/user-managment/roles.jsx]: actually role entity will be fetched in EntityTable component, so you can remove this, just fetch the permissions

File: src/main/webapp/js/utils/table-cells/collection-cell.jsx [hunk 58]
```diff
@@ -0,0 +1,30 @@
+import React from 'react';
+import { connect } from 'react-redux';
+import PropTypes from 'prop-types';
+
+import { getEntityArrayByName } from '../../selectors';
+
+const CollectionCell = ({ entityArray, relatedField }) => {
+  const collectionCellValue = JSON.stringify(entityArray[0][relatedField]);
```
Comment (pmuchowski) [inline src/main/webapp/js/utils/table-cells/collection-cell.jsx:8]: I think you should create a selector to get the list of values.
In this component use `_.join(values, ', ')` to display the values

**Review by pmuchowski — CHANGES_REQUESTED** [review]

File: src/main/java/org/motechproject/newebodac/domain/security/UserPermission.java [hunk 59]
```diff
@@ -14,7 +16,9 @@
 @Table(name = "user_permission")
 public class UserPermission extends BaseEntity {
 
-  @NotBlank
+public static UUID id;
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/newebodac/domain/security/UserPermission.java]: remove this

File: src/main/java/org/motechproject/newebodac/mapper/RoleMapper.java [hunk 60]
```diff
@@ -0,0 +1,33 @@
...
+  @Mapping(target = "updateDate", ignore = true)
+  @Mapping(target = "readonly", ignore = true)
+  void update(RoleDto roleDto, @MappingTarget UserRole role);
+
+  default UserPermission toUserPermission(UUID id) {
+    return new UserPermission(id);
+  }
+
+  default UUID toUuid(UserPermission userPermission) {
+    return UserPermission.id;
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/newebodac/mapper/RoleMapper.java]: use `userPermission.getId()`

File: src/main/webapp/js/components/navbar.jsx [hunk 61]
```diff
@@ -32,6 +33,7 @@ import MessageCampaign from './message-campaign';
 import Login from './auth/login';
 import ViewEntity from './entities/view-entity';
 import EditEntityPage from './entities/edit-entity-page';
+import ROLE_ENTITY from '../utils/entity-types';
```
Comment (pmuchowski) [inline src/main/webapp/js/components/navbar.jsx]: it's not default export, it will not work like this, add it in `{}`

**Review by pmuchowski — CHANGES_REQUESTED** [review]

**Pull request merged**